contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,381 @@
+        bytes32 group = _GROUP_OPEN;
+        _contractGroup[target] = group;
+        emit ContractGroupUpdated(target, group);
+    }
+
+    /**
+     * @dev Sets the target contract to be in the "closed" group. All restricted functions in the target contract will
+     * be closed down and disallowed to all. The caller must be the default admin.
+     */
+    function setContractClosed(address target) public virtual onlyDefaultAdmin {
+        bytes32 group = _GROUP_CLOSED;
+        _contractGroup[target] = group;
+        emit ContractGroupUpdated(target, group);
+    }
+
+    /**
+     * @dev Returns true if the team has already been created via {createTeam}.
+     */
+    function _teamExists(uint8 team) internal virtual returns (bool) {
+        return _getTeam(_createdTeams, team);
+    }
+
+    /**
+     * @dev Augmented version of {AccessControl-_grantRole} that keeps track of user team bitmaps.
+     */
+    function _grantRole(bytes32 role, address user) internal virtual override {
+        super._grantRole(role, user);
+        (bool isTeam, uint8 team) = _decodeTeamRole(role);
+        if (isTeam) {
+            require(_teamExists(team));
+            _userTeams[user] = _setTeam(_userTeams[user], team, true);
+        }
+    }
+
+    /**
+     * @dev Augmented version of {AccessControl-_revokeRole} that keeps track of user team bitmaps.
+     */
+    function _revokeRole(bytes32 role, address user) internal virtual override {
+        super._revokeRole(role, user);
+        (bool isTeam, uint8 team) = _decodeTeamRole(role);
+        if (isTeam) {
+            require(_teamExists(team));
+            require(team != _TEAM_PUBLIC);
+            _userTeams[user] = _setTeam(_userTeams[user], team, false);
+        }
+    }
+
+    /**
+     * @dev Returns the {AccessControl} role id that corresponds to a team.
+     *
+     * This role id starts with the ASCII characters `team:`, followed by zeroes, and ends with the single byte
+     * corresponding to the team number.
+     */
+    function _encodeTeamRole(uint8 team) internal virtual view returns (bytes32) {
+        return bytes32("team:") | bytes32(uint256(team));
+    }
+
+    /**
+     * @dev Decodes a role id into a team, if it is a role id of the kind returned by {_encodeTeamRole}.
+     */
+    function _decodeTeamRole(bytes32 role) internal virtual returns (bool, uint8) {
+        bytes32 tagMask = ~bytes32(uint256(0xff));
+        bytes32 tag = role & tagMask;
+        uint8 team  = uint8(role[31]);
+        return (tag == bytes32("team:"), team);
+    }
+
+    /**
+     * @dev Returns the isolate group id for a target contract.
+     *
+     * The group id consists of the ASCII characters `group:isolate:` followed by the contract address bytes.
+     */
+    function _encodeIsolateGroup(address target) internal virtual pure returns (bytes32) {
+        return _GROUP_ISOLATE_PREFIX | (bytes20(target) >> _GROUP_ISOLATE_PREFIX.length);
```

Comment:
I think you meant:

```suggestion
        return _GROUP_ISOLATE_PREFIX | (bytes32(bytes20(target)) >> _GROUP_ISOLATE_PREFIX.length * 8);
```

In which case, there are 2 bytes left out of the target address.

I wouldn't be worried about this since it's removing 2 bytes from the address collision security (20 bytes), reducing it to 2**18, which is still high enough. However, there might be some vanity addresses out there with multiple zeroes at the beginning.

According to a vanity address generator on an M1 Macbook Pro 16', getting a 7-zeros prefix is 5 months, reducing the security to 20 - 2 - 7 = 11 bytes. That starts to feel dangerously close of being non-negligible.

Moreover, some chains use predeploys or precompiles with hardcoded addresses that might imply a security concern if we trim the last two bytes. Consider the [Optimism predeploys](https://github.com/ethereum-optimism/optimism/blob/develop/specs/predeploys.md), where almost everyone starts with `0x42...` and the significant bits are the last two. 

Would be fixed if we replace the prefix with something shorter like `group:single` or `group:islte`.

Reply:
Oh interesting find. I definitely wouldn't want to erase any of the address bits so we should choose a shorter prefix.